channels/security/websocket.py

```diff
@@ -40,9 +54,88 @@ def valid_origin(self, origin):
         # None is not allowed
         if origin is None:
             return False
+        # Get only hostname all allowed origins
+        # Only if not full: https://domain.example.com:8443 -> domain.example.com
+        #   or //domain.example.com - > domain.example.com
+        #   or .example.com -> .example.com
+        self.allowed_origins = [urlparse(pattern).hostname or urlparse("//" + pattern).hostname
+                                or pattern for pattern in self.allowed_origins]
         # Check against our list
         return validate_host(origin, self.allowed_origins)
 
+    def valid_origin_full(self, origin):
+        # None is not allowed
+        if origin is None:
+            return False
+        # Check against our list
+        return self.validate_host_full(origin)
+
+    def validate_host_full(self, origin):
+        # Return True if valid, False otherwise
+        return any(pattern == "*" or self.is_allowed_origin(origin, pattern)
+                   for pattern in self.allowed_origins)
+
+    def is_allowed_origin(self, origin, pattern):
+        checked_address = 0
+        # Get ResultParse object
+        pattern = urlparse(pattern.lower())
+        # Get check result and origin ports
+        valid, origin_ports = self.check_pattern(origin, pattern)
+        if valid:
+            # Check ssl security certificate
+            if origin.scheme == "https" and self.check_cert:
+                for port in origin_ports:
+                    try:
+                        # Init HTTPSConnection object
+                        client = http_client.HTTPSConnection(origin.hostname, port)
+                        # Requesting a site with get method
+                        client.request("get", origin.hostname + origin.path)
+                        checked_address += 1
+                    # Except error: security certificate is invalid
+                    except (ssl.SSLError, ConnectionRefusedError) as error:
+                        # Check error
+                        if isinstance(error, ConnectionRefusedError):
+                            continue
+                        # Raise ssl.SSLError
+                        raise ssl.SSLError(origin.geturl() + " uses an invalid security certificate.")
+                if not checked_address:
+                    # Raise error, if all addresses are not available
+                    raise ConnectionRefusedError("Can not connect to at least one address.")
+            return True
+        return False
+
+    def get_origin_ports(self, origin):
+        if origin.port is not None:
+            # Return tuple (origin.port, )
+            return origin.port,
+        # if origin.port doesn`t exists
+        if origin.scheme == "http":
+            # Default ports return for http
+            return 80, 8080
```

Comment:
Is port 8080 actually specified as a default HTTP port anywhere?

Reply:
Yes